src/test/java/org/rulez/demokracia/pdengine/VoteTest.java

```diff
@@ -43,8 +42,98 @@ public void cast_vote_does_not_delete_ballot_if_canVote_is_false() {
+				 .assertMessageIs("This issue cannot be voted on yet");
+	}
+
+	@tested_feature("Vote")
+	@tested_operation("Cast vote")
+	@tested_behaviour("validates inputs")
+	@Test
+	public void cast_vote_checks_vote_id() {
+		String ballot = voteManager.obtainBallot(adminInfo.voteId, adminInfo.adminKey);
+		List<RankedChoice> theCastVote = new ArrayList<RankedChoice>();
+		Vote vote = getTheVote();
+		vote.canVote = true;
+		String wrongVoteId = "wrong_ID";
+
+		assertThrows(() -> voteManager.castVote(wrongVoteId, ballot, theCastVote)
+				).assertException(IllegalArgumentException.class)
+				 .assertMessageIs(String.format("illegal voteId: %s", wrongVoteId));
+	}
+
+	@tested_feature("Vote")
+	@tested_operation("Cast vote")
+	@tested_behaviour("validates inputs")
+	@Test
+	public void cast_vote_checks_ballot() {
+		String wrongBallot = RandomUtils.createRandomKey();
+		List<RankedChoice> theCastVote = new ArrayList<RankedChoice>();
+		Vote vote = getTheVote();
+		vote.canVote = true;
+
+		assertThrows(() -> voteManager.castVote(adminInfo.voteId, wrongBallot, theCastVote)
+				).assertException(IllegalArgumentException.class)
+				 .assertMessageIs(String.format("Illegal ballot: %s", wrongBallot));
+	}
+
+	@tested_feature("Vote")
+	@tested_operation("Cast vote")
+	@tested_behaviour("validates inputs")
+	@Test
+	public void cast_vote_checks_the_cast_if_choiceids_are_valid() {
+		String ballot = voteManager.obtainBallot(adminInfo.voteId, adminInfo.adminKey);
+		List<RankedChoice> theCastVote = new ArrayList<RankedChoice>();
+		Vote vote = getTheVote();
+		vote.canVote = true;	
+
+		String wrong_choiceId = RandomUtils.createRandomKey();
+		int good_rank = 42;
+		RankedChoice rankedChoice = new RankedChoice(wrong_choiceId, good_rank);
+		theCastVote.add(rankedChoice);
+
 		assertThrows(() -> voteManager.castVote(adminInfo.voteId, ballot, theCastVote)
-				).assertMessageIs("This issue cannot be voted on yet");
+				).assertException(IllegalArgumentException.class)
+				 .assertMessageIs(String.format("Invalid choiceId: %s", wrong_choiceId));
+	}
+
+	@tested_feature("Vote")
+	@tested_operation("Cast vote")
+	@tested_behaviour("validates inputs")
+	@Test
+	public void cast_vote_checks_the_cast_if_ranks_are_nonnegative() {
+		String ballot = voteManager.obtainBallot(adminInfo.voteId, adminInfo.adminKey);
+		List<RankedChoice> theCastVote = new ArrayList<RankedChoice>();
+		Vote vote = getTheVote();
+		vote.canVote = true;
+
+		String choiceId = vote.addChoice("valid_choice","userke");
+		int wrong_rank = -1;
+		RankedChoice rankedChoice = new RankedChoice(choiceId, wrong_rank);
+		theCastVote.add(rankedChoice);
+
+		assertThrows(() -> voteManager.castVote(adminInfo.voteId, ballot, theCastVote)
+				).assertException(IllegalArgumentException.class)
+				 .assertMessageIs(String.format("Invalid rank: %d", wrong_rank));
+	}
+	
+	@tested_feature("Vote")
+	@tested_operation("Cast vote")
+	@tested_behaviour("validates inputs")
```

Comment:
The tested behaviour here is
@tested_behaviour("deletes the ballot with ballotId, so only one vote is possible with a ballot")